File: web/lib/simulate.js
// Monte Carlo over the remaining qual matches, in the browser.
//
// Port of ranksim/simulate.py. Returns the exact shape the Python /api/simulate
// returned, so the rendering code did not have to change.
//
// The RNG differs from Python's Mersenne Twister, so a given seed does not
// reproduce Python's draws -- seeds are reproducible within this build only.
// The fit is what tests/test_parity.py pins exactly; the sampler is checked
// statistically.

const FORCE_MAX_TRIES = 60;
const TRUNCATE_MAX_TRIES = 16;

// mulberry32: small, fast, and seedable, which the browser's Math.random is not.
function mulberry32(seed) {
  let a = seed >>> 0;
  return function () {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function gaussianFactory(random) {
  let spare = null;
  return function gauss(mu, sigma) {
    if (spare !== null) {
      const value = mu + sigma * spare;
      spare = null;
      return value;
    }
    let u = 0;
    let v = 0;
    let s = 0;
    do {
      u = random() * 2 - 1;
      v = random() * 2 - 1;
      s = u * u + v * v;
    } while (s >= 1 || s === 0);
    const factor = Math.sqrt((-2 * Math.log(s)) / s);
    spare = v * factor;
    return mu + sigma * u * factor;
  };
}

function percentileFromCounts(counts, total, q) {
  const target = q * total;
  let seen = 0;
  for (let rank = 0; rank < counts.length; rank++) {
    seen += counts[rank];
    if (seen >= target) return rank + 1;
  }
  return counts.length;
}

function teamSortKey(team) {
  const digits = team.replace(/\D/g, "");
  return [digits ? Number(digits) : 0, team];
}

// The official tiebreaker chain: ranking score, then avg match, avg auto fuel,
// avg tower -- all averages over counting matches.
export function rankOrder(standings, teams) {
  return [...teams].sort((a, b) => {
    const ra = standings[a];
    const rb = standings[b];
    const na = ra.played || 1;
    const nb = rb.played || 1;
    const keys = [
      [rb.rp / nb, ra.rp / na],
      [rb.matchPoints / nb, ra.matchPoints / na],
      [rb.autoFuel / nb, ra.autoFuel / na],
      [rb.tower / nb, ra.tower / na],
    ];
    for (const [x, y] of keys) if (x !== y) return x - y;
    const [an, as] = teamSortKey(a);
    const [bn, bs] = teamSortKey(b);
    return an - bn || as.localeCompare(bs);
  });
}

export function simulate(bundle, fit, options = {}) {
  const {
    n = 5000,
    seed = null,
    oprUncertainty = true,
    cutoff = 8,
    forced = {},
    useScouting = true,
  } = options;

  const constants = bundle.constants;
  const teams = bundle.teams;
  const nTeams = teams.length;
  const index = {};
  teams.forEach((t, i) => (index[t] = i));

  const random = mulberry32(seed === null ? (Math.random() * 4294967296) >>> 0 : seed);
  const gauss = gaussianFactory(random);

  const base = bundle.standings;
  const baseRp = teams.map((t) => base[t].rp);
  const baseMp = teams.map((t) => base[t].matchPoints);
  const baseAf = teams.map((t) => base[t].autoFuel);
  const baseTw = teams.map((t) => base[t].tower);
  const basePlayed = teams.map((t) => base[t].played);

  const hubRating = teams.map((t) => fit.hub[t]);
  const hubSe = teams.map((t) => fit.hubSe[t]);
  const autoRating = teams.map((t) => fit.auto[t]);
  const autoSe = teams.map((t) => fit.autoSe[t]);

  // Per-team climb sampler: cumulative weights over point outcomes.
  const towerPoints = [];
  const towerCum = [];
  for (const t of teams) {
    const outcomes = fit.towerOutcomes[t];
    const pts = [];
    const cum = [];
    let running = 0;
    for (const [outcome, prob] of outcomes) {
      running += prob;
      pts.push(outcome.points);
      cum.push(running);
    }
    if (cum.length) cum[cum.length - 1] = 1;
    else {
      pts.push(0);
      cum.push(1);
    }
    towerPoints.push(pts);
    towerCum.push(cum);
  }

  const fouls = fit.foulSamples.length ? fit.foulSamples : [0];
  const thrEnergized = fit.thresholds.energized;
  const thrSupercharged = fit.thresholds.supercharged;
  const thrTraversal = fit.thresholds.traversal;
  const sigmaHub = fit.sigmaHub;
  const sigmaAuto = fit.sigmaAuto;
  const winRp = constants.winRp;
  const tieRp = constants.tieRp;

  const sched = bundle.remaining.map((m) => ({
    key: m.key,
    red: m.red.map((t) => index[t]),
    blue: m.blue.map((t) => index[t]),
    redCounting: m.redCounting.map((t) => index[t]),
    blueCounting: m.blueCounting.map((t) => index[t]),
    force: forced[m.key] || "",
  }));

  const rankCounts = teams.map(() => new Int32Array(nTeams));
  const rankSum = new Float64Array(nTeams);
  const rpSum = new Float64Array(nTeams);
  const rsSum = new Float64Array(nTeams);
  const matchCounts = {};
  for (const m of bundle.remaining) matchCounts[m.key] = { red: 0, blue: 0, tie: 0 };
  let forcedNudges = 0;

  // Integer draw from Normal(mean, sigma) truncated at zero. Clamping instead of
  // truncating piles an atom onto exactly 0, which shows up as spurious ties.
  const drawScore = (m, sigma) => {
    for (let i = 0; i < TRUNCATE_MAX_TRIES; i++) {
      const value = m + gauss(0, sigma);
      if (value >= 0) return Math.floor(value + 0.5);
    }
    return 0;
  };

  const drawTower = (idx) => {
    const r = random();
    const cum = towerCum[idx];
    for (let i = 0; i < cum.length; i++) if (r <= cum[i]) return towerPoints[idx][i];
    return towerPoints[idx][cum.length - 1];
  };

  const rp = new Float64Array(nTeams);
  const mp = new Float64Array(nTeams);
  const af = new Float64Array(nTeams);
  const tw = new Float64Array(nTeams);
  const played = new Float64Array(nTeams);
  const order = new Int32Array(nTeams);
  const hubR = new Float64Array(nTeams);
  const autoR = new Float64Array(nTeams);

  for (let trial = 0; trial < n; trial++) {
    for (let i = 0; i < nTeams; i++) {
      hubR[i] = oprUncertainty ? hubRating[i] + gauss(0, hubSe[i]) : hubRating[i];
      autoR[i] = oprUncertainty ? autoRating[i] + gauss(0, autoSe[i]) : autoRating[i];
      rp[i] = baseRp[i];
      mp[i] = baseMp[i];
      af[i] = baseAf[i];
      tw[i] = baseTw[i];
      played[i] = basePlayed[i];
    }

    for (const match of sched) {
      const { red, blue, force } = match;
      let rHub = 0;
      let bHub = 0;
      let rAuto = 0;
      let bAuto = 0;
      let rTower = 0;
      let bTower = 0;
      let rTotal = 0;
      let bTotal = 0;
      let landed = false;

      const tries = force === "red" || force === "blue" ? FORCE_MAX_TRIES : 1;
      for (let attempt = 0; attempt < tries; attempt++) {
        let rMean = 0;
        let bMean = 0;
        let rAutoMean = 0;
        let bAutoMean = 0;
        for (let i = 0; i < 3; i++) {
          rMean += hubR[red[i]];
          bMean += hubR[blue[i]];
          rAutoMean += autoR[red[i]];
          bAutoMean += autoR[blue[i]];
        }
        rHub = drawScore(rMean, sigmaHub);
        bHub = drawScore(bMean, sigmaHub);
        rAuto = drawScore(rAutoMean, sigmaAuto);
        bAuto = drawScore(bAutoMean, sigmaAuto);
        if (rAuto > rHub) rAuto = rHub;
        if (bAuto > bHub) bAuto = bHub;
        rTower = drawTower(red[0]) + drawTower(red[1]) + drawTower(red[2]);
        bTower = drawTower(blue[0]) + drawTower(blue[1]) + drawTower(blue[2]);
        rTotal = rHub + rTower + fouls[(random() * fouls.length) | 0];
        bTotal = bHub + bTower + fouls[(random() * fouls.length) | 0];
        if (!force || force === "tie") {
          landed = true;
          break;
        }
        if ((force === "red" && rTotal > bTotal) || (force === "blue" && bTotal > rTotal)) {
          landed = true;
          break;
        }
      }
      if (!landed && (force === "red" || force === "blue")) {
        // Never landed the requested upset; push the forced winner over the line
        // by the smallest amount that flips it.
        forcedNudges++;
        if (force === "red") {
          const bump = bTotal - rTotal + 1;
          rHub += bump;
          rTotal += bump;
        } else {
          const bump = rTotal - bTotal + 1;
          bHub += bump;
          bTotal += bump;
        }
      }

      let redWin;
      let blueWin;
      let tie;
      if (force === "tie") {
        redWin = false;
        blueWin = false;
        tie = true;
      } else {
        redWin = rTotal > bTotal;
        blueWin = bTotal > rTotal;
        tie = !redWin && !blueWin;
      }

      const counts = matchCounts[match.key];
      if (redWin) counts.red++;
      else if (blueWin) counts.blue++;
      else counts.tie++;

      let rRp = redWin ? winRp : tie ? tieRp : 0;
      let bRp = blueWin ? winRp : tie ? tieRp : 0;
      if (rHub >= thrEnergized) {
        rRp += 1;
        if (rHub >= thrSupercharged) rRp += 1;
      }
      if (bHub >= thrEnergized) {
        bRp += 1;
        if (bHub >= thrSupercharged) bRp += 1;
      }
      if (rTower >= thrTraversal) rRp += 1;
      if (bTower >= thrTraversal) bRp += 1;

      const rMp = rHub + rTower;
      const bMp = bHub + bTower;
      for (const i of match.redCounting) {
        rp[i] += rRp;
        mp[i] += rMp;
        af[i] += rAuto;
        tw[i] += rTower;
        played[i] += 1;
      }
      for (const i of match.blueCounting) {
        rp[i] += bRp;
        mp[i] += bMp;
        af[i] += bAuto;
        tw[i] += bTower;
        played[i] += 1;
      }
    }

    for (let i = 0; i < nTeams; i++) order[i] = i;
    order.sort((a, b) => {
      const pa = played[a] || 1;
      const pb = played[b] || 1;
      let d = rp[b] / pb - rp[a] / pa;
      if (d) return d;
      d = mp[b] / pb - mp[a] / pa;
      if (d) return d;
      d = af[b] / pb - af[a] / pa;
      if (d) return d;
      d = tw[b] / pb - tw[a] / pa;
      if (d) return d;
      return a - b;
    });

    for (let rank0 = 0; rank0 < nTeams; rank0++) {
      const i = order[rank0];
      rankCounts[i][rank0] += 1;
      rankSum[i] += rank0 + 1;
      rpSum[i] += rp[i];
      rsSum[i] += played[i] ? rp[i] / played[i] : 0;
    }
  }

  const clampedCutoff = Math.max(1, Math.min(cutoff, nTeams));
  const currentOrder = {};
  rankOrder(base, teams).forEach((t, i) => (currentOrder[t] = i + 1));

  const remainingByTeam = {};
  for (const m of bundle.remaining) {
    for (const t of m.redCounting.concat(m.blueCounting)) {
      (remainingByTeam[t] = remainingByTeam[t] || []).push(m.key);
    }
  }

  const teamRows = teams.map((team, i) => {
    const counts = rankCounts[i];
    const dist = Array.from(counts, (c) => c / n);
    let best = nTeams;
    let worst = 1;
    for (let r = 0; r < nTeams; r++) {
      if (counts[r]) {
        best = Math.min(best, r + 1);
        worst = Math.max(worst, r + 1);
      }
    }
    return {
      team,
      currentRank: currentOrder[team] || nTeams,
      current: base[team],
      meanRank: rankSum[i] / n,
      medianRank: percentileFromCounts(counts, n, 0.5),
      p05Rank: percentileFromCounts(counts, n, 0.05),
      p95Rank: percentileFromCounts(counts, n, 0.95),
      bestRank: best,
      worstRank: worst,
      pRank1: dist[0],
      pTop3: dist.slice(0, 3).reduce((a, b) => a + b, 0),
      pCutoff: dist.slice(0, clampedCutoff).reduce((a, b) => a + b, 0),
      expectedRp: rpSum[i] / n,
      expectedRankScore: rsSum[i] / n,
      rankDist: dist,
      remainingMatches: remainingByTeam[team] || [],
    };
  });
  teamRows.sort((a, b) => a.meanRank - b.meanRank || b.pCutoff - a.pCutoff);

  const matchRows = bundle.remaining.map((m) => {
    const c = matchCounts[m.key];
    return {
      key: m.key,
      number: m.number,
      red: m.red,
      blue: m.blue,
      redSurrogates: m.redSurrogates,
      blueSurrogates: m.blueSurrogates,
      played: false,
      winner: "",
      pRed: c.red / n,
      pBlue: c.blue / n,
      pTie: c.tie / n,
      forced: forced[m.key] || "",
      redPredicted: m.red.reduce((a, t) => a + fit.hub[t], 0),
      bluePredicted: m.blue.reduce((a, t) => a + fit.hub[t], 0),
    };
  });

  return {
    teams: teamRows,
    matches: matchRows,
    meta: {
      n,
      seed,
      cutoff: clampedCutoff,
      oprUncertainty,
      useScouting,
      forced,
      forcedNudged: forcedNudges,
      remainingMatches: bundle.remaining.length,
    },
  };
}

File: ranksim/simulate.py
"""Monte Carlo over the remaining qual matches.

One trial plays out every unplayed match, adds the result to the current
standings, re-sorts by the official tiebreaker chain, and records where each team
landed. Aggregating a few thousand of those gives the rank distribution.

Per trial the model draws:
  * a team rating vector (optionally resampled from each rating's standard error,
    so a team with two matches carries its uncertainty into the projection),
  * an alliance hub score around that rating, and its auto-fuel share,
  * a climb outcome per robot,
  * opponent-foul points, which decide close matches but never the tiebreakers.
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field

from .event import TIE_RP, WIN_RP, EventState, Match, TeamRecord, rank_teams
from .model import Fit

# How many resamples to spend trying to hit a forced winner before nudging the
# score. 60 tries covers anything down to a ~5% upset without distortion.
FORCE_MAX_TRIES = 60

# Alliance scores are drawn from a normal truncated at zero: the weakest
# alliances here have a predicted hub score below one residual sigma, so an
# unclamped normal puts real mass on negative scores, and clamping instead of
# truncating piles an atom onto exactly 0 -- which showed up as a spurious 1%+
# rate of 0-0 ties. Scores are then rounded, since real scores are integers and
# that is what makes a genuine tie possible at all.
TRUNCATE_MAX_TRIES = 16


@dataclass
class SimOptions:
    n: int = 5000
    seed: int | None = None
    opr_uncertainty: bool = True
    cutoff: int = 8
    forced: dict[str, str] = field(default_factory=dict)
    source: str = "csv"  # "csv" | "tba" -- which current standings to start from
    use_scouting: bool = True  # recorded for the caller; the fit already reflects it


def _base_records(state: EventState, source: str) -> dict[str, TeamRecord]:
    if source == "csv" and state.csv_records:
        base = {t: state.csv_records[t].copy() for t in state.csv_records}
        for team in state.teams:
            base.setdefault(team, TeamRecord(team))
        return base
    return {t: state.records[t].copy() for t in state.teams}


def _percentile_from_counts(counts: list[int], total: int, q: float) -> int:
    target = q * total
    seen = 0
    for rank, c in enumerate(counts, start=1):
        seen += c
        if seen >= target:
            return rank
    return len(counts)


def simulate(state: EventState, fit: Fit, options: SimOptions) -> dict:
    rng = random.Random(options.seed)
    gauss = rng.gauss

    teams = list(state.teams)
    n_teams = len(teams)
    index = {t: i for i, t in enumerate(teams)}

    base = _base_records(state, options.source)
    base_rp = [base[t].rp for t in teams]
    base_mp = [base[t].match_points for t in teams]
    base_af = [base[t].auto_fuel for t in teams]
    base_tw = [base[t].tower for t in teams]
    base_played = [base[t].played for t in teams]

    hub_rating = [fit.hub[t] for t in teams]
    hub_se = [fit.hub_se[t] for t in teams]
    auto_rating = [fit.auto[t] for t in teams]
    auto_se = [fit.auto_se[t] for t in teams]

    # Per-team climb sampler: cumulative weights over (points) outcomes.
    tower_points: list[list[int]] = []
    tower_cum: list[list[float]] = []
    for t in teams:
        outcomes = fit.tower_outcomes[t]
        pts, cum, running = [], [], 0.0
        for outcome, prob in outcomes:
            running += prob
            pts.append(outcome.points)
            cum.append(running)
        if cum:
            cum[-1] = 1.0
        else:
            pts, cum = [0], [1.0]
        tower_points.append(pts)
        tower_cum.append(cum)

    fouls = fit.foul_samples or [0]
    thr_energized = fit.thresholds["energized"]
    thr_supercharged = fit.thresholds["supercharged"]
    thr_traversal = fit.thresholds["traversal"]
    sigma_hub = fit.sigma_hub
    sigma_auto = fit.sigma_auto

    remaining: list[Match] = state.remaining
    # Precompute index lists so the hot loop touches no strings.
    sched = []
    for m in remaining:
        sched.append(
            (
                m.key,
                [index[t] for t in m.red],
                [index[t] for t in m.blue],
                [index[t] for t in m.counting("red")],
                [index[t] for t in m.counting("blue")],
                options.forced.get(m.key, ""),
            )
        )

    rank_counts = [[0] * n_teams for _ in range(n_teams)]
    rank_sum = [0.0] * n_teams
    rp_sum = [0.0] * n_teams
    rs_sum = [0.0] * n_teams
    match_counts = {m.key: {"red": 0, "blue": 0, "tie": 0} for m in remaining}
    forced_hits = 0
    forced_nudges = 0

    def draw_score(mean: float, sigma: float) -> int:
        """Integer draw from Normal(mean, sigma) truncated at zero."""
        for _ in range(TRUNCATE_MAX_TRIES):
            value = mean + gauss(0.0, sigma)
            if value >= 0.0:
                return int(value + 0.5)
        return 0

    def draw_tower(idx: int) -> int:
        r = rng.random()
        cum = tower_cum[idx]
        for i, c in enumerate(cum):
            if r <= c:
                return tower_points[idx][i]
        return tower_points[idx][-1]

    for _ in range(options.n):
        if options.opr_uncertainty:
            hub_r = [hub_rating[i] + gauss(0.0, hub_se[i]) for i in range(n_teams)]
            auto_r = [auto_rating[i] + gauss(0.0, auto_se[i]) for i in range(n_teams)]
        else:
            hub_r = hub_rating
            auto_r = auto_rating

        rp = base_rp[:]
        mp = base_mp[:]
        af = base_af[:]
        tw = base_tw[:]
        played = base_played[:]

        for key, red, blue, red_count, blue_count, force in sched:
            for _try in range(FORCE_MAX_TRIES if force in ("red", "blue") else 1):
                r_hub = draw_score(sum(hub_r[i] for i in red), sigma_hub)
                b_hub = draw_score(sum(hub_r[i] for i in blue), sigma_hub)
                r_auto = draw_score(sum(auto_r[i] for i in red), sigma_auto)
                b_auto = draw_score(sum(auto_r[i] for i in blue), sigma_auto)
                if r_auto > r_hub:
                    r_auto = r_hub
                if b_auto > b_hub:
                    b_auto = b_hub
                r_tower = draw_tower(red[0]) + draw_tower(red[1]) + draw_tower(red[2])
                b_tower = draw_tower(blue[0]) + draw_tower(blue[1]) + draw_tower(blue[2])
                r_total = r_hub + r_tower + rng.choice(fouls)
                b_total = b_hub + b_tower + rng.choice(fouls)
                if not force or force == "tie":
                    break
                if (force == "red" and r_total > b_total) or (
                    force == "blue" and b_total > r_total
                ):
                    forced_hits += 1
                    break
            else:
                # Never landed the requested upset; push the forced winner over
                # the line by the smallest amount that flips it.
                forced_nudges += 1
                if force == "red":
                    bump = b_total - r_total + 1
                    r_hub += bump
                    r_total += bump
                else:
                    bump = r_total - b_total + 1
                    b_hub += bump
                    b_total += bump

            if force == "tie":
                red_win = blue_win = False
                tie = True
            else:
                red_win = r_total > b_total
                blue_win = b_total > r_total
                tie = not (red_win or blue_win)

            counts = match_counts[key]
            counts["red" if red_win else "blue" if blue_win else "tie"] += 1

            r_rp = (WIN_RP if red_win else TIE_RP if tie else 0)
            b_rp = (WIN_RP if blue_win else TIE_RP if tie else 0)
            if r_hub >= thr_energized:
                r_rp += 1
                if r_hub >= thr_supercharged:
                    r_rp += 1
            if b_hub >= thr_energized:
                b_rp += 1
                if b_hub >= thr_supercharged:
                    b_rp += 1
            if r_tower >= thr_traversal:
                r_rp += 1
            if b_tower >= thr_traversal:
                b_rp += 1

            r_mp = r_hub + r_tower
            b_mp = b_hub + b_tower
            r_af = r_auto
            b_af = b_auto
            for i in red_count:
                rp[i] += r_rp
                mp[i] += r_mp
                af[i] += r_af
                tw[i] += r_tower
                played[i] += 1
            for i in blue_count:
                rp[i] += b_rp
                mp[i] += b_mp
                af[i] += b_af
                tw[i] += b_tower
                played[i] += 1

        order = sorted(
            range(n_teams),
            key=lambda i: (
                -rp[i] / played[i] if played[i] else 0.0,
                -mp[i] / played[i] if played[i] else 0.0,
                -af[i] / played[i] if played[i] else 0.0,
                -tw[i] / played[i] if played[i] else 0.0,
                i,
            ),
        )
        for rank0, i in enumerate(order):
            rank_counts[i][rank0] += 1
            rank_sum[i] += rank0 + 1
            rp_sum[i] += rp[i]
            rs_sum[i] += rp[i] / played[i] if played[i] else 0.0

    n = options.n
    cutoff = max(1, min(options.cutoff, n_teams))
    current_order = {t: i + 1 for i, t in enumerate(rank_teams(base))}

    team_rows = []
    for i, team in enumerate(teams):
        counts = rank_counts[i]
        dist = [c / n for c in counts]
        team_rows.append(
            {
                "team": team,
                "currentRank": current_order.get(team, n_teams),
                "current": base[team].as_dict(),
                "meanRank": rank_sum[i] / n,
                "medianRank": _percentile_from_counts(counts, n, 0.5),
                "p05Rank": _percentile_from_counts(counts, n, 0.05),
                "p95Rank": _percentile_from_counts(counts, n, 0.95),
                "bestRank": next(r + 1 for r, c in enumerate(counts) if c),
                "worstRank": next(
                    r + 1 for r in range(n_teams - 1, -1, -1) if counts[r]
                ),
                "pRank1": dist[0],
                "pTop3": sum(dist[:3]),
                "pCutoff": sum(dist[:cutoff]),
                "expectedRp": rp_sum[i] / n,
                "expectedRankScore": rs_sum[i] / n,
                "rankDist": dist,
                "remainingMatches": [
                    m.key for m in remaining if team in m.counting("red") + m.counting("blue")
                ],
            }
        )
    team_rows.sort(key=lambda r: (r["meanRank"], -r["pCutoff"]))

    match_rows = []
    for m in remaining:
        c = match_counts[m.key]
        match_rows.append(
            {
                **m.as_dict(),
                "pRed": c["red"] / n,
                "pBlue": c["blue"] / n,
                "pTie": c["tie"] / n,
                "forced": options.forced.get(m.key, ""),
                "redPredicted": sum(fit.hub[t] for t in m.red),
                "bluePredicted": sum(fit.hub[t] for t in m.blue),
            }
        )

    return {
        "teams": team_rows,
        "matches": match_rows,
        "meta": {
            "n": n,
            "seed": options.seed,
            "cutoff": cutoff,
            "oprUncertainty": options.opr_uncertainty,
            "source": options.source,
            "useScouting": options.use_scouting,
            "forced": options.forced,
            "forcedResampled": forced_hits,
            "forcedNudged": forced_nudges,
            "remainingMatches": len(remaining),
        },
    }
